File: internal/ai/claude.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type ClaudeEnricher struct {
	APIKey  string
	BaseURL string
	Model   string
	client  *http.Client
}

func NewClaudeEnricher(apiKey, baseURL string) *ClaudeEnricher {
	url := strings.TrimRight(baseURL, "/")
	if url == "" {
		url = "https://api.anthropic.com"
	}
	return &ClaudeEnricher{
		APIKey:  apiKey,
		BaseURL: url,
		Model:   "claude-sonnet-4-20250514",
		client:  &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *ClaudeEnricher) Name() string {
	return "claude"
}

func (c *ClaudeEnricher) Enrich(ctx context.Context, req EnrichRequest, categories []string) (EnrichResult, error) {
	if c.APIKey == "" {
		return EnrichResult{}, fmt.Errorf("Claude API key 未设置")
	}

	userMsg := fmt.Sprintf("File: %s, Version: %s, Extension: %s, Category: %s",
		req.Name, req.Version, req.Extension, req.Category)
	if len(categories) > 0 {
		userMsg += fmt.Sprintf(", Optional Functional Categories: %s", strings.Join(categories, ", "))
	}

	body := map[string]any{
		"model":      c.Model,
		"max_tokens": 1024,
		"system":     systemPrompt,
		"messages": []map[string]string{
			{"role": "user", "content": userMsg},
		},
	}
	jsonBody, _ := json.Marshal(body)

	httpReq, err := http.NewRequestWithContext(ctx, "POST", c.BaseURL+"/v1/messages", bytes.NewReader(jsonBody))
	if err != nil {
		return EnrichResult{}, err
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("x-api-key", c.APIKey)
	httpReq.Header.Set("anthropic-version", "2023-06-01")

	resp, err := c.client.Do(httpReq)
	if err != nil {
		return EnrichResult{}, fmt.Errorf("Claude API 调用失败: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return EnrichResult{}, err
	}

	if resp.StatusCode != http.StatusOK {
		return EnrichResult{}, fmt.Errorf("Claude API 返回错误 %d: %s", resp.StatusCode, string(respBody))
	}

	var msgResp struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.Unmarshal(respBody, &msgResp); err != nil {
		return EnrichResult{}, err
	}

	if len(msgResp.Content) == 0 {
		return EnrichResult{}, fmt.Errorf("Claude 返回空响应")
	}

	content := strings.TrimSpace(msgResp.Content[0].Text)
	content = strings.TrimPrefix(content, "```json")
	content = strings.TrimPrefix(content, "```")
	content = strings.TrimSuffix(content, "```")
	content = strings.TrimSpace(content)

	return ParseEnrichResponse([]byte(content), "claude")
}

File: src-tauri/src/ai/claude.rs
use std::future::Future;
use std::pin::Pin;
use std::time::Duration;

use crate::ai::enricher::{
    build_system_prompt, build_user_message, default_enrich_result, parse_enrich_response,
    EnrichRequest, EnrichResult, Enricher,
};

// ────────────────── Claude Enricher ──────────────────

pub struct ClaudeEnricher {
    api_key: String,
    base_url: String,
    model: String,
    client: reqwest::Client,
}

impl ClaudeEnricher {
    pub fn new(api_key: &str, base_url: &str) -> Self {
        Self {
            api_key: api_key.to_string(),
            base_url: if base_url.is_empty() {
                "https://api.anthropic.com".to_string()
            } else {
                base_url.to_string()
            },
            model: "claude-sonnet-4-20250514".to_string(),
            client: reqwest::Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .unwrap_or_default(),
        }
    }

    pub fn with_model(mut self, model: &str) -> Self {
        self.model = model.to_string();
        self
    }
}

impl Enricher for ClaudeEnricher {
    fn enrich(
        &self,
        req: EnrichRequest,
        categories: Vec<String>,
    ) -> Pin<Box<dyn Future<Output = Result<EnrichResult, Box<dyn std::error::Error + Send + Sync>>> + Send + '_>>
    {
        let api_key = self.api_key.clone();
        let base_url = self.base_url.clone();
        let model = self.model.clone();
        let client = self.client.clone();
        let name = req.name.clone();
        let tags = req.available_tags.clone().unwrap_or_default();
        let system_prompt = build_system_prompt(&categories, &tags);
        let user_msg = build_user_message(&req);

        Box::pin(async move {
            let url = format!("{}/v1/messages", base_url.trim_end_matches('/'));
            let body = serde_json::json!({
                "model": model,
                "max_tokens": 500,
                "system": system_prompt,
                "messages": [
                    { "role": "user", "content": user_msg },
                ],
            });

            let resp = client
                .post(&url)
                .header("x-api-key", &api_key)
                .header("anthropic-version", "2023-06-01")
                .header("Content-Type", "application/json")
                .json(&body)
                .send()
                .await
                .map_err(|e| -> Box<dyn std::error::Error + Send + Sync> {
                    format!("Claude request failed for '{}': {}", name, e).into()
                })?;

            let status = resp.status();
            if !status.is_success() {
                let status_text = resp.text().await.unwrap_or_default();
                let err_msg = format!("Claude {} error for '{}': {}", status, name, status_text);
                log::warn!("{}", err_msg);
                return Err(err_msg.into());
            }

            let json: serde_json::Value = resp.json().await?;
            // Claude 响应格式: { "content": [{ "type": "text", "text": "..." }] }
            let raw = json
                .pointer("/content/0/text")
                .and_then(|v| v.as_str())
                .ok_or_else(|| {
                    let err_msg = format!("Claude unexpected response format for '{}'", name);
                    log::warn!("{}", err_msg);
                    err_msg
                })?;

            Ok(parse_enrich_response(raw, "claude", &name))
        })
    }

    fn name(&self) -> &str {
        "claude"
    }
}
